### PR TITLE
Add product availability in channel

### PR DESCRIPTION
**Pull request by fowczarek**
- [x] Create a default channel in mutation if exists any Product or Order or Sale or Voucher or Checkout.
- [x] Create a new model for `ProductChannelListing`
- [x] Migrate publication data from `Product` to `ProductChannelListing`
- [x] Drop publication field from `Product`
- [x] Refactor `Product` querysets
- [x] Rafactor all availability usages in code not related to graphql 
- [x] Add protected queries for getting all `ProductChannelListing` for product
- [x] Refactor existing Product queries
- [x] Add warrining if query don't have `ChannelId` in request
- [x] Add mutation to manage `ProductChannelListing` 
- [x] Update fixtures in tests
- [ ] Update populate db - Separete PR 

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/graphql/product/types/products.py [hunk 1]
```diff
@@ -562,8 +528,12 @@ def calculate_pricing_with_collections(collections):
     @staticmethod
     def resolve_is_available(root: models.Product, info):
         country = info.context.country
+        channel_slug = str(info.context.channel_slug)
```
Comment (maarcingebala) [inline saleor/graphql/product/types/products.py:531]: Why `str()` here? In other places we don't do that.

File: saleor/graphql/translations/types.py [hunk 2]
```diff
@@ -124,12 +124,7 @@ class Meta:
 
     @staticmethod
     def resolve_product_variant(root: product_models.ProductVariant, info):
-        visible_products = product_models.Product.objects.visible_to_user(
-            info.context.user
-        ).values_list("pk", flat=True)
-        return product_models.ProductVariant.objects.filter(
-            product__id__in=visible_products, pk=root.id
-        ).first()
+        return product_models.ProductVariant.objects.filter(pk=root.id).first()
```
Comment (IKarbowiak) [inline saleor/graphql/translations/types.py:127]: If `root` is `ProductVariant` why we use filter and first here? Can't we just return root instance?

File: saleor/graphql/translations/types.py [hunk 3]
```diff
@@ -153,11 +148,7 @@ class Meta:
 
     @staticmethod
     def resolve_product(root: product_models.Product, info):
-        return (
-            product_models.Product.objects.visible_to_user(info.context.user)
-            .filter(pk=root.id)
-            .first()
-        )
+        return product_models.Product.objects.first()
```
Comment (IKarbowiak) [inline saleor/graphql/translations/types.py:151]: The same here. Just return `root`?

File: saleor/graphql/core/fields.py [hunk 4]
```diff
@@ -92,6 +92,16 @@ def convert_field_measurements(*_args):
 
 
 class PrefetchingConnectionField(BaseDjangoConnectionField):
+    def __init__(self, *args, **kwargs):
+        kwargs.setdefault(
+            "channel_slug",
+            graphene.Argument(
+                graphene.String,
+                description="Slug of the channel for which return data.",
```
Comment (maarcingebala) [inline saleor/graphql/core/fields.py:100]: Grammar police 🚔   

```suggestion
                description="Slug of the channel for which the data should be returned.",
```

File: saleor/graphql/core/fields.py [hunk 5]
```diff
@@ -213,3 +223,15 @@ def get_resolver(self, parent_resolver):
...
         )
+
+
+class FieldWithChannel(graphene.Field):
+    def __init__(self, type, *args, **kwargs):
+        kwargs.setdefault(
+            "channel_slug",
+            graphene.Argument(
+                graphene.String,
+                description="Slug of the channel for which return data.",
```
Comment (maarcingebala) [inline saleor/graphql/core/fields.py:234]:
```suggestion
                description="Slug of the channel for which the data should be returned.",
```

File: saleor/graphql/product/mutations/channels.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+if TYPE_CHECKING:
+    from ....product.models import Product as ProductModel
+    from ....channel.models import Channel as ChannelModel
+
+ErrorType = DefaultDict[str, List[ValidationError]]
+
+
+class ProductChannelListingAddInput(graphene.InputObjectType):
+    channel_id = graphene.ID(
+        required=True, description="ID of a channel witch be assigned to product."
```
Comment (maarcingebala) [inline saleor/graphql/product/mutations/channels.py:27]:
```suggestion
        required=True, description="ID of a channel."
```

File: saleor/graphql/product/mutations/channels.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+    class Arguments:
+        id = graphene.ID(required=True, description="ID of a product to update.")
+        input = ProductChannelListingUpdateInput(
+            required=True,
+            description="Fields required to create product channel listings.",
+        )
+
+    class Meta:
+        description = "Update product chanel listing."
```
Comment (maarcingebala) [inline saleor/graphql/product/mutations/channels.py:61]: The "channel listing" term is fine for mutations and internals but in the public description/docs I would explain more clearly that it refers to creating/updating product availability in a channel - this is how people should understand that.

```suggestion
        description = "Manage product's availability in channels."
```

File: saleor/graphql/product/mutations/channels.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+class ProductChannelListingUpdateInput(graphene.InputObjectType):
+    add_channels = graphene.List(
+        graphene.NonNull(ProductChannelListingAddInput),
+        description="List of channel assigned to product.",
+        required=False,
+    )
+    remove_channels = graphene.List(
+        graphene.NonNull(graphene.ID),
+        description="List of channel unassigned from product.",
```
Comment (maarcingebala) [inline saleor/graphql/product/mutations/channels.py:45]:
```suggestion
        description="List of channels from which the product should be unassigned.",
```

File: saleor/graphql/product/mutations/channels.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    )
+    publication_date = graphene.types.datetime.Date(
+        description="Publication date. ISO 8601 standard."
+    )
+
+
+class ProductChannelListingUpdateInput(graphene.InputObjectType):
+    add_channels = graphene.List(
+        graphene.NonNull(ProductChannelListingAddInput),
+        description="List of channel assigned to product.",
```
Comment (maarcingebala) [inline saleor/graphql/product/mutations/channels.py:40]:
```suggestion
        description="List of channels to which the product should be assigned.",
```

File: saleor/product/migrations/0121_productchannellisting.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+import django.db.models.deletion
+from django.db import migrations, models
+
+
+def migrate_products_publishable_data(apps, schema_editor):
+    Channel = apps.get_model("channel", "Channel")
+    Product = apps.get_model("product", "Product")
+    ProductChannelListing = apps.get_model("product", "ProductChannelListing")
+
+    if Product.objects.exists():
```
Comment (maarcingebala) [inline saleor/product/migrations/0121_productchannellisting.py:12]: I think this check is not needed, if there are no products then the iterator below would not run. One query less #performance ;)

**Pull request merged**